service/controller/clusterapi/v26/key/provider.go

```diff
@@ -0,0 +1,57 @@
+package key
+
+import (
+	"encoding/json"
+
+	g8sv1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/cluster/v1alpha1"
+	"k8s.io/apimachinery/pkg/runtime"
+	cmav1alpha1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+)
+
+func mustG8sSpecFromCMASpec(cmaSpec cmav1alpha1.ProviderSpec) g8sv1alpha1.AWSClusterSpec {
+	if cmaSpec.Value == nil {
+		panic("provider spec value must not be empty")
+	}
+
+	var g8sSpec g8sv1alpha1.AWSClusterSpec
+	{
+		if len(cmaSpec.Value.Raw) == 0 {
+			return g8sSpec
+		}
+
+		err := json.Unmarshal(cmaSpec.Value.Raw, &g8sSpec)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	return g8sSpec
+}
+
+func mustG8sStatusFromCMAStatus(cmaStatus *runtime.RawExtension) g8sv1alpha1.AWSClusterStatus {
+	if cmaStatus == nil {
+		panic("provider status value must not be empty")
```

Comment:
I see the function name and I agree that these are useful right now when developing, but should we maybe change this when heading further towards production? I can imagine someone creating fishy CRs and when we have a crashing operator stuck.

Reply:
We can see how to go about this later on. My thinking was that the key functions should not return errors. I would rather like to have a validating admission controller to ensure the provider config is set properly.

Reply:
I sort of agree that errors in key functions are not very nice, but with all my experience on clusters where single dev can crash the whole ingress controller of a cluster and therefore break the traffic is just unacceptable and therefore I'd like to build the operator so that single CR object doesn't crash the operator.

Lets see how to go here. For now I'm fine with this.